ingress.yaml

```diff
@@ -1,3 +1,36 @@
+v3:
+  requires:
+    type: object
+    properties:
+      service:
+        type: string
+      port:
+        type: integer
+      namespace:
+        type: string
+      prefix:
+        type: string
+      rewrite:
+        type: string
+      per_unit_routes:
+        type: boolean
+    required:
+      - service
+      - port
+      - namespace
+      - prefix
+  provides:
```

Comment:
I'm confused as to why `istio-pilot` would be providing `url` and `unit_urls` back to the `requires` side of the relation. In V3 of this relation, wouldn't the `prometheus` charm want to just say "here's multiple URLs and the endpoints to be routed to for each" instead of "here's one URL and the endpoint to be routed to"?

Reply:
You've got it a bit backwards. This is for an app, such as kubernetes-worker, to tell prometheus how to connect to it so that prometheus will scrape / pull in the metrics. And given that, (AFAIK) prometheus doesn't use app-level data with the leader sending a list of URLs for prometheus to scrape but instead relies on each unit to advertise its own ingress URL an thus each unit needs to be able to identify which URL belongs to it. (My original approach was, in fact, to just return a list of all ingress URLs but this change was specifically requested.)

Reply:
Unless I'm misunderstanding something, that sounds like it should be a different relation?

`ingress` is a somewhat nebulous term, but the way this relation was designed was handling a charm requiring being routed to by a reverse proxy or similar. In other words, it's built to handle `Relation B` in this diagram, but it seems like you want `Relation A`?

```
                  ┌─────────────────────────────┐
                  │           Cluster           │
┌────────────┐    │ ┌─────────┐     ┌─────────┐ │
│            │    │ │         │     │         │ │
│ Prometheus ├────┼─► Ingress ├─────► Service │ │
│            │  ▲ │ │         │  ▲  │         │ │
└────────────┘  │ │ └─────────┘  │  └─────────┘ │
                │ │              │              │
                │ └──────────────┼──────────────┘
                │                │
                │                │
            Relation A        Relation B
```

(On a semi-related note, wow https://asciiflow.com/ is nifty :smile:)

Reply:
That can't be the relation pattern, though.  It represents the traffic flow, but the relations are meant to model the logical connections between applications used to provide or consume services. `Prometheus` doesn't care if `Service` is fronted by a proxy, gateway, LB or whatever; it just needs to know that it should be connected to, and thus scrape metrics from `Service` and to be told what URLs to use to scrape said metrics. Likewise, `Ingress` doesn't care what the routable addresses / URLs it provides are used for or who they're given out to, it's job is to have ingress requests sent to it, create routable addresses / URLs, and respond to the request with those addresses / URLs. If it were to sit between `Service` and `Prometheus`, or any other combination of applications using ingress, then it would need to grow support for every conceivable interface type that might want to use those ingress addresses and somehow know how to provide what addresses to which relation endpoints, which it simply could not do.  FWIW, the "sit between two services" approach is the same anti-pattern that was originally used by the kubeapi-load-balancer charm which really only "worked" because it was hyper-focused to a single use-case and even then lead to the relation configuration being wildly different depending for slightly different usage patterns and led to a lot of bugs with deployments when those weren't set up correctly.

Reply:
@johnsca I think scraping ingress routing is a separate concern/relation from regular ingress routing. For example, `jupyter-ui` needs ingress routing, but `jupyter-controller` doesn't. They both however, probably want scraping ingress routing. Also, `jupyter-ui` has a different port for metrics/scraping than its regular service port, so there'd have to be a separate relation anyways. Could we do something like this?

```
          `public-url`
                │
                │  ┌─────────────┐
                │  │   Cluster   │
┌────────────┐  │  │ ┌─────────┐ │
│            │  ▼  │ │         │ │
│ Prometheus─┼─────┼─► Ingress │ │
│            │     │ │         │ │
└──────▲─────┘     │ └────▲────┘ │
       │           │      │◄───────── `scrape-ingress`
       │           │ ┌────┴────┐ │
       │           │ │         │ │
       └───────────┼─┤ Service │ │
          ▲        │ │         │ │
          │        │ └─────────┘ │
 `scrape-target`   │             │
                   └─────────────┘
```

Where `ingress` isn't shown because it's not relevant for prometheus or scraping in general. `public-url` would be how prometheus knows how to connect to the cluster. `scrape-ingress` would look similar to how `ingress` looks, except allow for multiple routes to be declared, to support the per-unit scraping. `scrape-target` would then simply present a list of paths such as `['/my/scrape/path/unit-123', '/my/scrape/path/unit-456']`, and prometheus would just scrape `<public-url>/my/path/unit-123`.

Possibly, `scrape-ingress` and `scrape-target` could be collapsed into a single relation, but point is, I think the `ingress` relation is unrelated to prometheus scraping.

Reply:
Regardless of of whether the ingress a charm requests is going to be used for a dashboard UI, metrics scraping, a data pipeline, or whatever else is immaterial to the nature of the interface and it doesn't make any sense to duplicate the interface for every potential use-case. Being caught up on the specific use-case of Prometheus is counterproductive; this applies equally well to any client & server situation.

Adding a `public-url` relation endpoint that has to be additionally attached means that you complicate *every* charm that includes a host / port or URL value in any of its relations, just on the off chance that it might need to support the other side being connected via an ingress controller.  That's completely unscalable and untenable. If a client and server charm are connected together via a relation so that they can know to talk to each other, the relation for that should not change whether the communication happens to be within the same model, cross-model on the same controller, cross-controller on the same cluster, or to a controller on the other side of the internet.

You're right that a given charm will quite possibly need multiple network endpoints exposed via ingress, possibly with different properties, and this schema doesn't support that as currently defined. I mentioned this limitation [in another comment](https://github.com/canonical/istio-operators/pull/43#discussion_r766110019).  I guess both you and Michele are envisioning it as the charm having a separate relation endpoints for each service, while I am envisioning it as one relation endpoint which uses the request / response pattern to get ingress for everything it needs. The separate relation endpoint approach does allow for better configurability and control over things like having an actual public ingress controller vs a cross-internal-zone ingress controller, but it also complicates the model by exposing a bunch of internal architecture and can't handle something like dynamic services. But adding support for multiple requests doesn't preclude a charm from defining multiple relation endpoints, and in fact I did this very thing w/ the (very similar) `loadbalancer` interface in the CK charms.

At the end of the day, the purpose of the `ingress` interface is fundamentally about a charm asking an ingress controller "Please translate this network endpoint into one that can be reached from elsewhere." The ingress provider should not be involved beyond that, and whatever the charm does with translated endpoint is up to it. (Also, I was purposefully vague about "network endpoint" because this really shouldn't be limited to only HTTP resources; we have need to expose application network endpoints as well and there's no reason this interface shouldn't just handle both.)

Reply:
I advocate for a `public-url` relation, because one already needs to exist: `dex-auth` and `oidc-gatekeeper` would benefit from one, we just haven't gotten around to implementing it yet. They both require a manual `juju config charm public-url=http://foo/`.

`ingress` is different from `scrape-ingress` (which is not particular to prometheus, it could be any service that does metrics scraping) already: metrics scraping doesn't need `rewrite` for example, because it's never serving html/javascript that refers back to itself. It also won't need things that eventually will crop up for `ingress`, such as `allowed-request-headers`/`allowed-response-headers`, `timeout`, or `x-forwarded-prefix`, all of which exist now in the upstream Kubeflow manifests repository. `ingress` is also different in that per-unit routing is a peculiarity of metrics scraping, and not something that any other service wants or needs.

Reply:
I don't understand why you think you need an additional `public-url` relation?  You already have a relation which semantically means "give me a public entry point to this service" and it only makes sense for that relation to return the created public entry point / URL so that the requesting side doesn't have to guess.  Where does the need for an additional relation come in?

And just because a use-case doesn't need all of the features of an interface doesn't make it semantically any different. But I also disagree that metrics wouldn't need any of the features you mentioned; their use-cases might be uncommon, but that's no reason to proscribe them. The same goes for per-unit routing: it may not be a widely-used use-case, but just because you can't think of another reason why it might be needed, that doesn't mean it warrants forking off the entire interface.

You have to think about interfaces as semantic markers and you also need to keep in mind the trade-offs between making them more specific vs more general. The more you make interfaces tailored to a single use-case, the more relation endpoints charms have to grow, and that growth is multiplicative over the combination of number of charms and number of interfaces. Obviously, an interface can become too general to the point that it loses semantic meaning, but "ingress" seems pretty clearly defined as "give me a publicly accessible point of entry for this thing" and whether that thing is the app-level service or a single unit, whether it uses advanced header features or not, it is still semantically the same request.

Reply:
Just had a good read of this thread. I think the tl;dr is I'm leaning more toward @johnsca's implementation here.

I can definitely appreciate that having a `public-url` relation would be more explicit/expressive in the model, which in general is something I'd like to encourage, but I think  in this case it might just be *too* specific, and seen as extra toil required when establishing the relations needed for an app to be exposed, whether that be for scraping or otherwise. While Prometheus is the obvious example here, 

I'm sure there will be other cases where having support for routing to specific units will be necessary, so I see no need to have two separate relation interfaces to support that. I think it's reasonable that such a use-case might be a little more complex to set up, because in reality it *is* more of an edge case, but either way I don't think maintaining a whole other interface is the way forward.